src/frequenz/client/dispatch/test/_service.py

```diff
@@ -178,6 +179,44 @@ def _filter_dispatch(
                 if dispatch.dry_run != _filter.is_dry_run:
                     return False
 
+            if len(_filter.dispatch_ids) > 0 and dispatch.id not in map(
+                DispatchId, _filter.dispatch_ids
+            ):
+                return False
+
+            if not FakeService.matches_query(_filter, dispatch):
+                return False
+
+        return True
+
+    @staticmethod
+    def matches_query(_filter: DispatchFilter, dispatch: Dispatch) -> bool:
+        """Check if a dispatch matches the query."""
+        # Two cases:
+        #  - query starts with # and is interpretable as int: filter by id
+        #  - otherwise: filter by exact match in 'type' field
+        # Multiple queries use OR logic - dispatch must match at least one
+        if len(_filter.queries) > 0:
+            matches_any_query = False
+            for query in _filter.queries:
+                if query.startswith("#"):
+                    try:
+                        query_id = DispatchId(int(query[1:]))
+                        if dispatch.id == query_id:
+                            matches_any_query = True
+                            break
+                    except ValueError:
+                        # not an int, interpret as exact type match (without the #)
+                        if query[1:] == dispatch.type:
+                            matches_any_query = True
+                            break
```

Comment:
To make the error detection more fine-grained:


```suggestion
                    try:
                        int_id = int(query[1:])
                    except ValueError:
                        # not an int, interpret as exact type match (without the #)
                        if query[1:] == dispatch.type:
                            matches_any_query = True
                            break
                    else:
                        query_id = DispatchId(int_id)
                        if dispatch.id == query_id:
                            matches_any_query = True
                            break
```